tests/framework/builder.py

```diff
@@ -40,11 +56,99 @@ def build(self, kernel: Artifact, disks: [DiskArtifact], config: Artifact):
         # Apply the microvm artifact configuration
         vm.basic_config(vcpu_count=int(microvm_config['vcpu_count']),
                         mem_size_mib=int(microvm_config['mem_size_mib']),
-                        ht_enabled=bool(microvm_config['ht_enabled']))
-
+                        ht_enabled=bool(microvm_config['ht_enabled']),
+                        track_dirty_pages=enable_diff_snapshots,
+                        boot_args='console=ttyS0 reboot=k panic=1')
         return vm
 
-    # TBD: Snapshot support is not fully merged. Once that is read
-    # this function will provide a way to spin up clones.
-    def build_from_snapshot(self, snapshot: Artifact):
+    def build_from_snapshot(self,
+                            snapshot: Snapshot,
+                            host_ip,
+                            guest_ip,
+                            netmask_len,
+                            resume=False,
+                            # Enable incremental snapshot capability.
+                            enable_diff_snapshots=False):
         """Build a microvm from a snapshot artifact."""
+        self.init_root_path()
+        vm = init_microvm(self.root_path, self.bin_cloner_path)
+
+        # We are going to use screen.
+        vm.jailer.daemonize = False
+        vm.jailer.extra_args.update({'seccomp-level': '0'})
+        vm.spawn(create_logger=False)
+
+        # Setup logging path.
+        log_fifo_path = os.path.join(vm.path, 'firecracker.log')
+        log_fifo = log_tools.Fifo(log_fifo_path)
+
+        # Configure logging.
+        response = vm.logger.put(
+            log_path=vm.create_jailed_resource(log_fifo.path),
+            level='Debug',
+            show_level=True,
+            show_log_origin=True,
+        )
+
+        assert vm.api_session.is_status_no_content(response.status_code)
+        # Hardlink all the snapshot files into the microvm jail.
+        jailed_mem = vm.create_jailed_resource(snapshot.mem)
+        jailed_vmstate = vm.create_jailed_resource(snapshot.vmstate)
+        _jailed_disk = vm.create_jailed_resource(snapshot.disks[0])
```

Comment:
Can `snapshot.disks == []`?

Reply:
Why not add all the disks in the jailer?

Reply:
There is a check for disks being non in the Snapshot Constructor. It is possible to boot a microvm without a disk so it can be an empty array. However there is not such test yet.

Reply:
Added assertion as suggested by @iulianbarbu .